src/rust/generator-executor/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+    #[structopt(env)]
+    plugin_work_queue_address: String,
+    // Path to the certificate to be used when validating the generator plugin TLS sessions
+    #[structopt(env)]
+    generator_certificate_path: std::path::PathBuf,
+    #[structopt(flatten)]
+    generator_client_config: GeneratorClientConfig,
+}
+
+fn make_job_stream(
+    mut work_queue_client: PluginWorkQueueServiceClient<tonic::transport::Channel>,
+) -> impl Stream<Item = (ExecutionJob, i64)> {
+    stream! {
+        loop {
+            let message = match work_queue_client.get_execute_generator(GetExecuteGeneratorRequest {}).await {
+                Ok(message) => message,
+                Err(e) => {
+                    tracing::error!(message="Failed to retrieve generator job", error=?e);
+                    // check what kind of error it is and retry it accordingly
+                    sleep(Duration::from_secs(1)).await;
+                    continue
+                }
+            };
+            match message {
+                GetExecuteGeneratorResponse { execution_job: Some(job), request_id } => yield (job, request_id),
+                _ => {
+                    tracing::debug!(message="No available jobs");
+                    sleep(Duration::from_secs(1)).await;
+                }
+            };
+
+        }
+    }
+}
+
+// Attempts to ack the message. On failure, will retry in a background task
+async fn try_ack(
+    work_queue_client: &mut PluginWorkQueueServiceClient<tonic::transport::Channel>,
+    request_id: i64,
+    success: bool,
+) {
+    if let Err(e) = work_queue_client
+        .acknowledge_generator(AcknowledgeGeneratorRequest {
+            request_id,
+            success,
+        })
+        .timeout(Duration::from_secs(2))
+        .await
+    {
+        let mut work_queue_client = work_queue_client.clone();
+        tokio::spawn(async move {
+            tracing::error!(
+                message="Failed to acknowledge message after 2 seconds",
+                error=?e,
+            );
+            if let Err(e) = work_queue_client
+                .acknowledge_generator(AcknowledgeGeneratorRequest {
+                    request_id,
+                    success,
+                })
+                .timeout(Duration::from_secs(10))
+                .await
+            {
+                tracing::error!(
+                    message="Failed to acknowledge message after 10 seconds",
+                    error=?e,
+                )
+            }
+        });
+    };
+}
+
+pub async fn process_message_loop(
+    mut work_queue_client: PluginWorkQueueServiceClient<tonic::transport::Channel>,
+    mut generator_client: GeneratorClient,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let jobs = make_job_stream(work_queue_client.clone());
+    pin_mut!(jobs);
+
+    while let Some((job, request_id)) = jobs.next().await {
+        let ExecutionJob {
+            tenant_id: _,
+            plugin_id,
+            data,
+        } = job;
```

Comment:
If you don't care about `tenant_id`, I think you can just do:
```rust
let ExecutionJob { plugin_id, data, ... } = job;
```

Reply:
If in fact we don't care about the tenant_id it should be removed from the protobuf also. The semantics of an "optional" tenant ID are pretty concerning from a safety standpoint, and `plugin_id` should be "unique enough" I think?